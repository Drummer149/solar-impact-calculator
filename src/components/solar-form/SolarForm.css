.MuiFormControl-root {
  width: 100%;
  margin-top: 10px !important;
}

#submitBtn {
  width: 50%;
  color: white;
  background-color: #05AC72;
  margin-top: 15px;
}