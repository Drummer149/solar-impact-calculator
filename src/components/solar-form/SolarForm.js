import React, { Component } from 'react';
import { DatePicker, MuiPickersUtilsProvider } from "@material-ui/pickers";
import MomentUtils from '@date-io/moment';
import { TextField, Button } from '@material-ui/core';
import './SolarForm.css';
import Moment from 'moment';
import PropTypes from 'prop-types';
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator';
 
/**
 * Componet of a form for a user to input their solor specifications
 */
class SolarForm extends Component {
  constructor(props){
    super(props);

    this.state = {
      startDate: Moment.now(),
      currentInstallation: 0,
    };

    // Bind functions to the this for form usage
    this.onChange = this.onChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  // Push form data to the state
  onChange(e){
    this.setState({ [e.target.name]: e.target.value });
  }

  // Push date change to the state
  onDateChange(e){
    this.setState({ startDate: e });
  }

  // Take the state data and send it to parent
  onSubmit(e){
    e.preventDefault();
    const installationData = {
      startDate: this.state.startDate,
      currentInstallation: this.state.currentInstallation,
    };

    this.props.setSolarInstallationSpecs(installationData);
  }

  render() {
    return (
      <React.Fragment>
        <h4>How switching to Solar has affected the world?</h4>
        <ValidatorForm onSubmit={this.onSubmit}>
        <MuiPickersUtilsProvider utils={MomentUtils}>
          <DatePicker
            label="When was your solar system installed?"
            name="startDate"
            format="MMMM Do YYYY"
            value={this.state.startDate}
            onChange={this.onDateChange}
            maxDate={new Date()}
            />
        </ MuiPickersUtilsProvider>
        <TextValidator 
          type="number"
          placeholder="How many Panels were installed"
          label="How many panels were installed?"
          name="currentInstallation"
          validators={['minNumber:0', 'required']}
          errorMessages={['Must be a positive number', 'This field is required']}
          onChange={this.onChange}
          value={this.state.currentInstallation}
          min="0"
          />
          <Button id="submitBtn" type="submit">
            Submit
          </Button>
          <p>Based on Heckert 310 Wp high performance modules,<br /> in comparison to Coal fired power production</p>
        </ValidatorForm>
      </React.Fragment>
    )
  }
}

export default SolarForm;