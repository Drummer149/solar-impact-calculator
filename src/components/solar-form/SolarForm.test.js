import React from 'react';
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import SolarForm from './SolarForm';
import Moment from 'moment';

Enzyme.configure({adapter: new Adapter()});

describe('<SolarForm />', () => {

  const setInstalltionMock = jest.fn();
  const preventDefaultMock = jest.fn();

  const defaultProps = {
    setSolarInstallationSpecs: setInstalltionMock
  };

  const event = {
    preventDefault: preventDefaultMock,
    target: {
      name: 'currentInstallation',
      value: 10
    }
  };

  const dateNow = Moment.now();
  
  test('Renders', () => {
    const wrapper = shallow(<SolarForm {...defaultProps} />);
    expect(wrapper.exists()).toBe(true);
  })

  test('Renders correctly', () => {
    const wrapper = shallow(<SolarForm {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  })

  test('Component has a method on change and will expect the state to change given an HTMLInputEvent', () => {
    const wrapper = shallow(<SolarForm {...defaultProps} />);
    wrapper.instance().onChange(event)
    expect(wrapper.state().currentInstallation).toBe(event.target.value);
  })

  test('Component has a method on date change and will expect the state to change given a Moment Date', () => {
    const wrapper = shallow(<SolarForm {...defaultProps} />);
    wrapper.instance().onDateChange(dateNow);
    expect(wrapper.state().startDate).toBe(dateNow);
  })

  test('Component has a method on submit change and will expect the state to change given a Moment Date', () => {
    const wrapper = shallow(<SolarForm {...defaultProps} />);
    wrapper.instance().onChange(event);
    wrapper.instance().onDateChange(dateNow);
    wrapper.instance().onSubmit(event);

    let mockData = {
      startDate: dateNow,
      currentInstallation: event.target.value,
    }

    expect(setInstalltionMock).toBeCalledWith(mockData);
  })
});
