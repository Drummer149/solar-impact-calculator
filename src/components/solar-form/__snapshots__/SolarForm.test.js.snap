// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SolarForm /> Renders correctly 1`] = `ShallowWrapper {}`;
